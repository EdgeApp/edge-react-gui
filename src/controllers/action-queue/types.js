// @flow
/* eslint-disable no-use-before-define */

//
// Action Operations
//
export type ActionOpTypes =
  | 'seq'
  | 'par'
  | 'exchange-buy'
  | 'exchange-sell'
  | 'loan-borrow'
  | 'loan-deposit'
  | 'loan-repay'
  | 'loan-withdraw'
  | 'swap'
  | 'toast'
  | 'delay'

export type ActionOpExecStatus = 'pending' | 'active' | 'done' | Error

export type ActionOp =
  | {
      type: 'seq',
      actions: ActionOp[]
    }
  | {
      type: 'par',
      actions: ActionOp[]
    }
  | {
      type: 'broadcast-tx',
      pluginId: string,
      rawTx: Uint8Array
    }
  | {
      type: 'exchange-buy',
      nativeAmount: string,
      walletId: string,
      tokenId?: string,
      exchangePluginId: string
    }
  | {
      type: 'exchange-sell',
      nativeAmount: string,
      walletId: string,
      tokenId?: string,
      exchangePluginId: string
    }
  | {
      type: 'loan-borrow',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-deposit',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-repay',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-withdraw',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'swap',
      fromWalletId: string,
      toWalletId: string,
      fromTokenId?: string,
      toTokenId?: string,
      nativeAmount: string,
      amountFor: 'from' | 'to'
    }
  // Useful for development/testing
  | {
      type: 'toast',
      message: string
    }
  | {
      type: 'delay',
      ms: number
    }

//
// Action (After) Effects
//

export type ActionEffect =
  | {
      type: 'seq',
      opIndex: number,
      childEffect: ActionEffect
    }
  | {
      type: 'par',
      childEffects: ActionEffect[]
    }
  | {
      type: 'address-balance',
      address: string,
      aboveAmount?: string,
      belowAmount?: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'tx-confs',
      txId: string,
      walletId: string,
      confirmations: number
    }
  | {
      type: 'price-level',
      currencyPair: string,
      aboveRate?: number,
      belowRate?: number
    }
  | {
      type: 'done',
      error?: Error
    }
  // Useful for development/testing
  | {
      type: 'unixtime',
      timestamp: number
    }
  | {
      type: 'init'
    }

//
// Action Program
//

// On Disk:
export type ActionProgram = {
  programId: string,
  actionOp: ActionOp
}

export type ActionProgramState = {
  programId: string,
  effect?: ActionEffect
}

//
// Internal Types
//

export type ActionQueueItem = {
  program: ActionProgram,
  state: ActionProgramState
}
export type ActionQueueMap = {
  [id: string]: ActionQueueItem
}

export type ExecutionResult = {
  effect: ActionEffect,
  action?: ActionOp // For dryrun
}
export type ExecutionResults = {
  nextState: ActionProgramState
}

//
// Aciton Display API
//

export type ActionDisplayInfo = {
  title: string,
  message: string,
  status: ActionOpExecStatus,
  steps: ActionDisplayInfo[]
}
