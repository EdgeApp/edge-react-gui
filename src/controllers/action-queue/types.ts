/* eslint-disable no-use-before-define */

import { EdgeAccount, EdgeNetworkFee, EdgeTransaction } from 'edge-core-js'

//
// Action Operations
//
export type ActionOpTypes =
  | 'seq'
  | 'par'
  | 'wyre-buy'
  | 'wyre-sell'
  | 'loan-borrow'
  | 'loan-deposit'
  | 'loan-repay'
  | 'loan-withdraw'
  | 'swap'
  | 'toast'
  | 'delay'

export type SeqActionOp = {
  type: 'seq'
  actions: ActionOp[]
}
export type ParActionOp = {
  type: 'par'
  actions: ActionOp[]
}
export type BroadcastTxActionOp = {
  type: 'broadcast-tx'
  pluginId: string
  rawTx: Uint8Array
}
export type WyreBuyActionOp = {
  type: 'wyre-buy'
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type WyreSellActionOp = {
  type: 'wyre-sell'
  wyreAccountId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanBorrowActionOp = {
  type: 'loan-borrow'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanDepositActionOp = {
  type: 'loan-deposit'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanRepayActionOp = {
  type: 'loan-repay'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanWithdrawActionOp = {
  type: 'loan-withdraw'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type SwapActionOp = {
  type: 'swap'
  amountFor: 'from' | 'to'
  fromTokenId?: string
  fromWalletId: string
  nativeAmount: string
  toTokenId?: string
  toWalletId: string
}
export type ActionOp =
  | SeqActionOp
  | ParActionOp
  | BroadcastTxActionOp
  | WyreBuyActionOp
  | WyreSellActionOp
  | LoanBorrowActionOp
  | LoanDepositActionOp
  | LoanRepayActionOp
  | LoanWithdrawActionOp
  | SwapActionOp

//
// Action (After) Effects
//

export type SeqEffect = {
  type: 'seq'
  opIndex: number
  childEffects: Array<ActionEffect | null> // null is only for dryrun
}
export type ParEffect = {
  type: 'par'
  childEffects: Array<ActionEffect | null> // null is only for dryrun
}
export type AddressBalanceEffect = {
  type: 'address-balance'
  address: string
  aboveAmount?: string
  belowAmount?: string
  walletId: string
  tokenId?: string
}
export type PushEventEffect = {
  type: 'push-event'
  eventId: string
}
export type PriceLevelEffect = {
  type: 'price-level'
  currencyPair: string
  aboveRate?: number
  belowRate?: number
}
export type TxConfsEffect = {
  type: 'tx-confs'
  txId: string
  walletId: string
  confirmations: number
}
export type DoneEffect = {
  type: 'done'
  error?: Error
  cancelled?: boolean
}

export type ActionEffect = SeqEffect | ParEffect | AddressBalanceEffect | PushEventEffect | PriceLevelEffect | TxConfsEffect | DoneEffect

//
// Action Program
//

// Storage:
export type ActionProgram = {
  programId: string
  actionOp: ActionOp
  // Development mode flag
  mockMode?: boolean
}
export type ActionProgramState = {
  clientId: string
  programId: string
  effect?: ActionEffect

  // Flags:
  effective: boolean // Whether the effect is observed
  executing: boolean // Whether the program is executing

  lastExecutionTime: number // The time when the effect was checked
  nextExecutionTime: number // The next time when the effect should be checked again
}

export type ActionQueueItem = {
  program: ActionProgram
  state: ActionProgramState
}
export type ActionQueueMap = {
  [id: string]: ActionQueueItem
}

// Runtime:
export type BroadcastTx = {
  walletId: string
  networkFee: EdgeNetworkFee
  tx: EdgeTransaction
}
export type ExecutableAction = {
  dryrunOutput: ExecutionOutput | null
  execute: () => Promise<ExecutionOutput>
}
export type ExecutionContext = {
  account: EdgeAccount
  clientId: string
}
export type ExecutionOutput = {
  effect: ActionEffect
  broadcastTxs: BroadcastTx[]
}
export type ExecutionResults = {
  nextState: ActionProgramState
}
export type PendingTxMap = {
  [walletId: string]: EdgeTransaction[]
}

//
// Aciton Display API
//

export type ActionDisplayStatus = 'pending' | 'active' | 'done' | Error

export type ActionDisplayInfo = {
  title: string
  message: string
  status: ActionDisplayStatus
  steps: ActionDisplayInfo[]
}
