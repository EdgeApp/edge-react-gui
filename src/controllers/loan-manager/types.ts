import { BorrowEngine, BorrowPlugin } from '../../plugins/borrow-plugins/types'
import { LoanProgramEdge } from './store'

export type LoanAccount = {
  id: string
  borrowPlugin: BorrowPlugin
  borrowEngine: BorrowEngine

  // Action Program Relationships
  programEdges: LoanProgramEdge[]
}

export type LoanAccountMap = {
  [walletId: string]: LoanAccount
}

export type LoanManager = {
  loanAccounts: LoanAccount[]
}
