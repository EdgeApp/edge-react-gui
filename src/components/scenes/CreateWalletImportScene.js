// @flow

import { type EdgeAccount } from 'edge-core-js'
import * as React from 'react'
import { ActivityIndicator, StyleSheet, View } from 'react-native'

import { CURRENCY_PLUGIN_NAMES, getSpecialCurrencyInfo } from '../../constants/WalletAndCurrencyConstants.js'
import s from '../../locales/strings.js'
import { PrimaryButton } from '../../modules/UI/components/Buttons/PrimaryButton.ui.js'
import { FormattedText as Text } from '../../modules/UI/components/FormattedText/FormattedText.ui.js'
import { Gradient } from '../../modules/UI/components/Gradient/Gradient.ui'
import { SafeAreaViewComponent as SafeAreaView } from '../../modules/UI/components/SafeAreaView/SafeAreaView.ui.js'
import { THEME } from '../../theme/variables/airbitz.js'
import { PLATFORM } from '../../theme/variables/platform.js'
import { connect } from '../../types/reactRedux.js'
import { type NavigationProp, type RouteProp } from '../../types/routerTypes.js'
import { scale } from '../../util/scaling.js'
import { FormField } from '../common/FormField.js'
import { ButtonsModal } from '../modals/ButtonsModal.js'
import { Airship } from '../services/AirshipInstance.js'

type OwnProps = {
  navigation: NavigationProp<'createWalletImport'>,
  route: RouteProp<'createWalletImport'>
}
type StateProps = {
  account: EdgeAccount
}
type Props = OwnProps & StateProps

type State = {
  input: string,
  error: string,
  isProcessing: boolean,
  cleanedPrivateKey: string
}

class CreateWalletImportComponent extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = {
      input: '',
      error: '',
      isProcessing: false,
      cleanedPrivateKey: ''
    }
  }

  handleNext = (): void => {
    const { account, navigation, route } = this.props
    const { selectedWalletType } = route.params
    const { input } = this.state
    const { currencyCode, walletType } = selectedWalletType
    const currencyPluginName = CURRENCY_PLUGIN_NAMES[currencyCode]
    const currencyPlugin = account.currencyConfig[currencyPluginName]
    const [, format] = walletType.split('-')

    this.setState({ isProcessing: true })
    currencyPlugin
      .importKey(input, { keyOptions: { format } })
      .then(() => {
        navigation.navigate('createWalletSelectFiat', {
          selectedWalletType,
          cleanedPrivateKey: input
        })
      })
      .catch(error =>
        Airship.show(bridge => (
          <ButtonsModal
            bridge={bridge}
            buttons={{ ok: { label: s.strings.string_ok } }}
            message={error.message}
            title={s.strings.create_wallet_failed_import_header}
          />
        ))
      )
      .then(() => this.setState({ isProcessing: false }))
  }

  onChangeText = (input: string) => {
    this.setState({ input })
  }

  render() {
    const { route } = this.props
    const { error, isProcessing, input } = this.state
    const { selectedWalletType } = route.params
    const { currencyCode } = selectedWalletType

    const specialCurrencyInfo = getSpecialCurrencyInfo(currencyCode)
    if (!specialCurrencyInfo.isImportKeySupported) throw new Error()
    const instructionSyntax = specialCurrencyInfo.isImportKeySupported.privateKeyInstructions
    const labelKeySyntax = specialCurrencyInfo.isImportKeySupported.privateKeyLabel
    return (
      <SafeAreaView>
        <View style={styles.scene}>
          <Gradient style={styles.gradient} />
          <View style={styles.view}>
            <View style={styles.createWalletPromptArea}>
              <Text style={styles.instructionalText}>{instructionSyntax}</Text>
            </View>
            <FormField
              autoFocus
              autoCorrect={false}
              onChangeText={this.onChangeText}
              label={labelKeySyntax}
              value={input}
              returnKeyType="next"
              onSubmitEditing={this.handleNext}
              multiline
              error={error}
            />
            <View style={styles.buttons}>
              <PrimaryButton style={styles.next} onPress={this.handleNext} disabled={isProcessing}>
                {isProcessing ? (
                  <ActivityIndicator color={THEME.COLORS.ACCENT_MINT} />
                ) : (
                  <PrimaryButton.Text>{s.strings.string_next_capitalized}</PrimaryButton.Text>
                )}
              </PrimaryButton>
            </View>
          </View>
        </View>
      </SafeAreaView>
    )
  }
}

const rawStyles = {
  scene: {
    flex: 1,
    backgroundColor: THEME.COLORS.WHITE
  },
  gradient: {
    height: THEME.HEADER,
    width: '100%',
    position: 'absolute'
  },
  view: {
    position: 'relative',
    top: THEME.HEADER,
    paddingHorizontal: 20,
    height: PLATFORM.usableHeight
  },
  createWalletPromptArea: {
    paddingTop: scale(16),
    paddingBottom: scale(8)
  },
  instructionalText: {
    fontSize: scale(16),
    textAlign: 'center',
    color: THEME.COLORS.GRAY_1
  },
  buttons: {
    marginTop: scale(24),
    flexDirection: 'row'
  },
  next: {
    marginLeft: scale(1),
    flex: 1
  }
}
const styles: typeof rawStyles = StyleSheet.create(rawStyles)

export const CreateWalletImportScene = connect<StateProps, {}, OwnProps>(
  state => ({
    account: state.core.account
  }),
  dispatch => ({})
)(CreateWalletImportComponent)
