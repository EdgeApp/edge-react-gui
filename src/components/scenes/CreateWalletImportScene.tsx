import { JsonObject } from 'edge-core-js'
import * as React from 'react'
import { View } from 'react-native'
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view'
import { sprintf } from 'sprintf-js'

import { PLACEHOLDER_WALLET_ID, splitCreateWalletItems } from '../../actions/CreateWalletActions'
import ImportKeySvg from '../../assets/images/import-key-icon.svg'
import { SPECIAL_CURRENCY_INFO } from '../../constants/WalletAndCurrencyConstants'
import { useHandler } from '../../hooks/useHandler'
import { lstrings } from '../../locales/strings'
import { useSelector } from '../../types/reactRedux'
import { EdgeSceneProps } from '../../types/routerTypes'
import { SceneWrapper } from '../common/SceneWrapper'
import { ButtonsModal } from '../modals/ButtonsModal'
import { Airship } from '../services/AirshipInstance'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { EdgeText } from '../themed/EdgeText'
import { MainButton } from '../themed/MainButton'
import { OutlinedTextInput, OutlinedTextInputRef } from '../themed/OutlinedTextInput'
import { SceneHeader } from '../themed/SceneHeader'
import { WalletCreateItem } from '../themed/WalletList'

export interface CreateWalletImportParams {
  createWalletList: WalletCreateItem[]
  walletNames: { [key: string]: string }
  fiatCode: string
}

interface Props extends EdgeSceneProps<'createWalletImport'> {}

const CreateWalletImportComponent = (props: Props) => {
  const { navigation, route } = props
  const { createWalletList, walletNames, fiatCode } = route.params
  const theme = useTheme()
  const styles = getStyles(theme)

  const account = useSelector(state => state.core.account)
  const { currencyConfig } = account

  const [importText, setImportText] = React.useState('')
  const [scrollEnabled, setScrollEnabled] = React.useState(false)

  const textInputRef = React.useRef<OutlinedTextInputRef>(null)

  const handleNext = useHandler(async () => {
    textInputRef.current?.blur()
    const cleanImportText = importText.trim()

    // Test imports
    const { newWalletItems } = splitCreateWalletItems(createWalletList)

    const pluginIds = newWalletItems.map(item => item.pluginId)

    // Loop over plugin importPrivateKey
    const promises = pluginIds.map(
      async pluginId =>
        await currencyConfig[pluginId].importKey(cleanImportText).catch(e => {
          console.warn('importKey failed', e)
        })
    )

    const results = await Promise.all(promises)

    const successMap: { [pluginId: string]: JsonObject } = {}

    for (const [i, keys] of results.entries()) {
      if (typeof keys === 'object') {
        // Success
        successMap[pluginIds[i]] = keys
      }
    }

    // Split up the original list of create items into success and failure lists
    const failureItems: WalletCreateItem[] = []
    const successItems: WalletCreateItem[] = []

    for (const item of createWalletList) {
      if (successMap[item.pluginId] != null) {
        // Any asset associated to this pluginId is good to go
        successItems.push(item)
      } else if (item.createWalletIds != null && item.createWalletIds[0] === PLACEHOLDER_WALLET_ID) {
        // Token items to be enabled on existing wallets and aren't dependent on a failed import are are good to go, too
        successItems.push(item)
      } else {
        // No good
        failureItems.push(item)
      }
    }

    if (successItems.length === 0) {
      await Airship.show<'edit' | undefined>(bridge => (
        <ButtonsModal
          bridge={bridge}
          title={lstrings.create_wallet_failed_import_header}
          message={lstrings.create_wallet_all_failed}
          buttons={{
            edit: { label: lstrings.create_wallet_edit }
          }}
        />
      ))

      return
    }

    if (failureItems.length > 0) {
      // Show modal with errors
      const displayNames = failureItems.map(item => item.displayName).join(', ')
      const resolveValue = await Airship.show<'continue' | 'edit' | 'cancel' | undefined>(bridge => (
        <ButtonsModal
          bridge={bridge}
          title={lstrings.create_wallet_failed_import_header}
          message={sprintf(lstrings.create_wallet_some_failed, displayNames)}
          buttons={{
            continue: { label: lstrings.legacy_address_modal_continue },
            cancel: { label: lstrings.string_cancel_cap }
          }}
        />
      ))

      if (resolveValue === 'cancel' || resolveValue == null) {
        return
      }
    }

    if (pluginIds.length > 0 && pluginIds.some(pluginId => SPECIAL_CURRENCY_INFO[pluginId]?.importKeyOptions != null)) {
      navigation.navigate('createWalletImportOptions', { createWalletList, walletNames, fiatCode, importText: cleanImportText })
      return
    }
    navigation.navigate('createWalletCompletion', { createWalletList, walletNames, fiatCode, importText: cleanImportText })
  })

  // Scale the icon
  const svgHeight = React.useMemo(() => 36 * theme.rem(0.0625), [theme])
  const svgWidth = React.useMemo(() => 83 * theme.rem(0.0625), [theme])

  return (
    <SceneWrapper background="theme">
      <SceneHeader title={lstrings.create_wallet_import_title} withTopMargin />
      <KeyboardAwareScrollView
        contentContainerStyle={styles.container}
        scrollEnabled={scrollEnabled}
        keyboardShouldPersistTaps="handled"
        onKeyboardDidChangeFrame={() => setScrollEnabled(false)}
        onKeyboardWillChangeFrame={() => setScrollEnabled(true)}
      >
        <View style={styles.icon}>
          <ImportKeySvg accessibilityHint={lstrings.import_key_icon_hint} color={theme.iconTappable} height={svgHeight} width={svgWidth} />
        </View>
        <EdgeText style={styles.instructionalText} numberOfLines={2}>
          {lstrings.create_wallet_import_all_instructions}
        </EdgeText>
        <OutlinedTextInput
          value={importText}
          returnKeyType="next"
          label={lstrings.create_wallet_import_input_key_or_seed_prompt}
          autoCapitalize="none"
          autoCorrect={false}
          blurOnClear={false}
          onChangeText={setImportText}
          onSubmitEditing={handleNext}
          marginRem={[1, 0.75, 1.25]}
          ref={textInputRef}
        />
        <MainButton label={lstrings.string_next_capitalized} type="secondary" marginRem={[0.5, 0.5]} onPress={handleNext} alignSelf="center" />
      </KeyboardAwareScrollView>
    </SceneWrapper>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    marginTop: theme.rem(0.5)
  },
  icon: {
    flexDirection: 'row',
    justifyContent: 'center',
    marginVertical: theme.rem(2)
  },
  instructionalText: {
    fontSize: theme.rem(1),
    color: theme.primaryText,
    paddingHorizontal: theme.rem(1),
    marginTop: theme.rem(0.5),
    marginBottom: theme.rem(1),
    marginHorizontal: theme.rem(0.5),
    textAlign: 'center'
  }
}))

export const CreateWalletImportScene = React.memo(CreateWalletImportComponent)
