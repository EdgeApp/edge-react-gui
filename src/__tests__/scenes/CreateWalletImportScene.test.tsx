import { describe, expect, it, jest } from '@jest/globals'
import * as React from 'react'
import TestRenderer from 'react-test-renderer'

import { CreateWalletImportScene } from '../../components/scenes/CreateWalletImportScene'
import { RouteProp } from '../../types/routerTypes'
import { fakeNavigation } from '../../util/fake/fakeNavigation'
import { fakeNonce } from '../../util/fake/fakeNonce'
import { FakeProviders, FakeState } from '../../util/fake/FakeProviders'

jest.mock('react-native-keyboard-aware-scroll-view', () => {
  const KeyboardAwareScrollView = (blob: { children: React.ReactNode }) => blob.children
  return { KeyboardAwareScrollView }
})

// Jest doesn't like direct SVG imports
jest.mock('../../assets/images/import-key-icon.svg', () => 'ImportKeySvg')

describe('CreateWalletImportScene', () => {
  const nonce = fakeNonce(0)
  const mockState: FakeState = {
    core: {
      account: {
        currencyConfig: {
          bitcoin: {
            importKey: () => {}
          }
        }
      }
    }
  }

  it('should render with loading props', () => {
    const navigation = fakeNavigation
    const route: RouteProp<'createWalletImport'> = {
      key: `createWalletImport-${nonce()}`,
      name: 'createWalletImport',
      params: {
        createWalletList: [
          {
            key: `create-wallet:bitcoin-bip49-bitcoin`,
            currencyCode: 'BTC',
            displayName: 'Bitcoin',
            pluginId: 'bitcoin',
            walletType: 'wallet:bitcoin-bip49'
          }
        ],
        walletNames: { 'create-wallet:bitcoin-bip49-bitcoin': 'My Bitcoin' },
        fiatCode: 'USD'
      }
    }

    const renderer = TestRenderer.create(
      <FakeProviders initialState={mockState}>
        <CreateWalletImportScene navigation={navigation} route={route} />
      </FakeProviders>
    )

    expect(renderer.toJSON()).toMatchSnapshot()
  })
})
