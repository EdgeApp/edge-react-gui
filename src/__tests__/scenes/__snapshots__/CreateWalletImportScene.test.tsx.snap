// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateWalletImportScene should render with loading props 1`] = `
<BVLinearGradient
  colors={
    [
      4278585612,
      4279971643,
    ]
  }
  endPoint={
    {
      "x": 1,
      "y": 0,
    }
  }
  locations={null}
  startPoint={
    {
      "x": 0,
      "y": 0,
    }
  }
  style={
    {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "stretch",
          "flexDirection": "column",
          "justifyContent": "flex-start",
          "position": "absolute",
        },
        {
          "bottom": 0,
          "left": 0,
          "padding": 0,
          "right": 0,
          "top": 64,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "justifyContent": "center",
            "marginHorizontal": 22,
            "paddingBottom": 22,
          },
          {
            "marginTop": 22,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.65}
          numberOfLines={1}
          style={
            [
              {
                "color": "#FFFFFF",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "includeFontPadding": false,
              },
              {
                "fontFamily": "Quicksand-Medium",
                "fontSize": 27,
              },
              null,
            ]
          }
        >
          Import Wallet
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "center",
          "marginVertical": 45,
        }
      }
    >
      <ImportKeySvg
        accessibilityHint="Import Key"
        color="#00f1a2"
        height={36}
        width={83}
      />
    </View>
    <Text
      adjustsFontSizeToFit={true}
      minimumFontScale={0.65}
      numberOfLines={2}
      style={
        [
          {
            "color": "#FFFFFF",
            "fontFamily": "Quicksand-Regular",
            "fontSize": 22,
            "includeFontPadding": false,
          },
          {
            "color": "#FFFFFF",
            "fontSize": 22,
            "marginBottom": 22,
            "marginHorizontal": 11,
            "marginTop": 11,
            "paddingHorizontal": 22,
            "textAlign": "center",
          },
          null,
        ]
      }
    >
      Enter your private seed, private key, or active key to verify and restore the associated wallet
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={false}
      focusable={true}
      onClick={[Function]}
      onLayout={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "rgba(255, 255, 255, 0)",
            "borderRadius": 11,
            "flexDirection": "row",
            "minHeight": 67,
            "paddingHorizontal": 22,
          },
          {
            "flexGrow": 0,
            "marginBottom": 28,
            "marginLeft": 17,
            "marginRight": 17,
            "marginTop": 22,
            "paddingLeft": 22,
            "paddingRight": 65,
            "paddingVertical": 0,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "borderTopWidth": 1,
              "bottom": 0,
              "left": 22,
              "position": "absolute",
              "right": 22,
            },
            {
              "borderColor": undefined,
              "right": 11,
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "borderBottomLeftRadius": 11,
              "borderBottomWidth": 1,
              "borderLeftWidth": 1,
              "borderRightWidth": 0,
              "borderTopLeftRadius": 11,
              "borderTopWidth": 1,
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "top": 0,
              "width": 22,
            },
            {
              "borderColor": undefined,
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "borderBottomRightRadius": 11,
              "borderBottomWidth": 1,
              "borderLeftWidth": 0,
              "borderRightWidth": 1,
              "borderTopRightRadius": 11,
              "borderTopWidth": 1,
              "bottom": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 22,
            },
            {
              "borderColor": undefined,
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "borderTopWidth": 1,
              "left": 22,
              "position": "absolute",
              "right": 22,
              "top": 0,
            },
            {
              "borderColor": undefined,
              "left": 22,
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "height": 67,
              "justifyContent": "center",
              "left": 0,
              "paddingHorizontal": 22,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            {
              "paddingLeft": 22,
              "paddingRight": 65,
            },
          ]
        }
      >
        <Text
          accessible={true}
          numberOfLines={1}
          onLayout={[Function]}
          style={
            [
              {
                "alignSelf": "flex-start",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 22,
                "padding": 0,
              },
              {
                "color": undefined,
                "transform": [
                  {
                    "translateX": 0,
                  },
                  {
                    "translateY": -0,
                  },
                  {
                    "scale": 1,
                  },
                ],
              },
            ]
          }
          testID="undefined.labelText"
        >
          Private Key or Private Seed
        </Text>
      </View>
      <View
        style={
          {
            "bottom": -45,
            "height": 45,
            "left": 0,
            "paddingLeft": 22,
            "paddingRight": 39,
            "position": "absolute",
            "right": 0,
          }
        }
      >
        <Text
          accessible={true}
          numberOfLines={2}
          style={
            [
              {
                "color": "#E85466",
                "fontFamily": "Quicksand-Regular",
                "fontSize": 17,
              },
              {
                "opacity": 0,
              },
            ]
          }
          testID="undefined.subText"
        />
      </View>
      <Text
        accessible={true}
        numberOfLines={1}
        onLayout={[Function]}
        style={
          [
            {
              "bottom": -10,
              "fontFamily": "Quicksand-Regular",
              "fontSize": 17,
              "position": "absolute",
              "right": 28,
            },
            {
              "color": undefined,
              "opacity": 0,
              "transform": [
                {
                  "translateX": 6,
                },
                {
                  "scale": 0,
                },
              ],
            },
          ]
        }
        testID="undefined.charLimit"
      />
      <TextInput
        accessibilityState={
          {
            "disabled": false,
          }
        }
        accessible={true}
        autoCapitalize="none"
        autoCorrect={false}
        autoFocus={true}
        editable={true}
        multiline={false}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        returnKeyType="next"
        secureTextEntry={false}
        selectionColor="#FFFFFF"
        style={
          [
            {
              "color": "#FFFFFF",
              "flexGrow": 1,
              "flexShrink": 1,
              "fontFamily": "Quicksand-Regular",
              "padding": 0,
            },
            {
              "fontSize": 22,
            },
          ]
        }
        testID="undefined.textInput"
        textAlignVertical="top"
        value=""
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "elevation": 0,
          "opacity": 1,
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
          "shadowRadius": 0,
        }
      }
    >
      <BVLinearGradient
        colors={
          [
            16777215,
            16777215,
          ]
        }
        endPoint={
          {
            "x": 1,
            "y": 1,
          }
        }
        locations={null}
        startPoint={
          {
            "x": 0,
            "y": 0,
          }
        }
        style={
          [
            {
              "alignItems": "center",
              "borderColor": "#00f1a2",
              "borderRadius": 6,
              "borderWidth": 1.5,
              "flexDirection": "row",
              "justifyContent": "center",
              "minHeight": 67,
              "minWidth": 202,
            },
            {
              "alignSelf": "center",
              "marginBottom": 11,
              "marginLeft": 11,
              "marginRight": 11,
              "marginTop": 11,
              "opacity": 1,
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
            },
            {
              "borderRadius": 6,
              "paddingLeft": 0,
              "paddingRight": 0,
            },
          ]
        }
      >
        <Text
          adjustsFontSizeToFit={true}
          minimumFontScale={0.75}
          numberOfLines={1}
          style={
            {
              "color": "#00f1a2",
              "fontFamily": "Quicksand-Regular",
              "fontSize": 22,
              "marginHorizontal": 0,
              "paddingBottom": 11,
              "paddingLeft": 17,
              "paddingRight": 17,
              "paddingTop": 11,
            }
          }
        >
          Next
        </Text>
      </BVLinearGradient>
    </View>
  </View>
</BVLinearGradient>
`;
